### PR TITLE
[FIRRTL][LowerXMR] Lower internal path into a module

### PR DESCRIPTION
**Pull request by prithayan**
This commit adds support to generate XMR with opaque internal paths into a module.
A combination of `verbatim.expr` and `ref.send` is used to encode an internal path into a module. `LowerXMR` must generate the correct hierarchical path for such internal paths into a module.
If an internal path (for example `random.path.inside`) exists inside a module `Foo`, the hierarchical path must end at the op that instantiates `Foo`, and the `random.path.inside` must be appended to it. This avoids any temporary `NodeOp` or wires inside the module `Foo`.
This replaces the requirement for a new op as discussed in https://github.com/llvm/circt/pull/4196

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp [hunk 1]
```diff
@@ -64,6 +64,20 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
...
+                  // generating the correct XMR, no node can be created in this
+                  // module. Create a null InnerRef and ensure the hierarchical
+                  // path ends at the parent that instantiates this module.
+                  auto inRef = InnerRefAttr();
+                  auto ind = addReachingSendsEntry(send.getResult(), inRef);
+                  xmrPathSuffix[ind] = ("." + verbExpr.getText()).str();
+                  markForRemoval(verbExpr);
+                  markForRemoval(send);
+                  return success();
+                }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:80]: What should happen here if the condition doesn't hold? Or: is there a reason to gate this based on the symbol being empty? I think it's just a reason to not mark the `verbExpr` for removal.

File: lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp [hunk 2]
```diff
@@ -64,6 +64,20 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
...
+              if (auto verbExpr = dyn_cast<VerbatimExprOp>(xmrDefOp))
+                if (verbExpr.getSymbolsAttr().empty() &&
+                    xmrDefOp->hasOneUse()) {
+                  // This represents the internal path into a module. For
+                  // generating the correct XMR, no node can be created in this
+                  // module. Create a null InnerRef and ensure the hierarchical
+                  // path ends at the parent that instantiates this module.
+                  auto inRef = InnerRefAttr();
+                  auto ind = addReachingSendsEntry(send.getResult(), inRef);
+                  xmrPathSuffix[ind] = ("." + verbExpr.getText()).str();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:76]: Is there a situation where the PR shouldn't add a `"."` leading prefix?

**Review by seldridge — APPROVED** [review]
Generally looks fine. Couple of questions on the PR.

File: lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp (hunk 1 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:80]: And also, we currently are not handling the symbol substitution for suffix. It should be simple to update the data structures to keep track of the symbols also. 
So, the condition is narrowly checking for a verbatim with an opaque "internal path", that would be generated by the `internalPath` annotation. 
You are right, if I keep track of the symbols for suffix substitution, the check is not required, other than marking for removal.

File: lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp (hunk 2 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp:76]: You are right, we always need it. So, it can potentially be moved to the use of the suffix.

**Pull request merged**